File: kranji-reading/kranji-reading-app/src/main/java/kranji/reading/app/zi/ZiSelectionSecretaryModule.java
package kranji.reading.app.zi;

import hue.captains.singapura.js.homing.core.DomModule;
import hue.captains.singapura.js.homing.core.Exportable;
import hue.captains.singapura.js.homing.core.ExportsOf;
import hue.captains.singapura.js.homing.core.ImportsFor;

import java.util.List;

/**
 * The Secretary for the character-selection Party.
 *
 * <p>A second bus, deliberately separate from navigation. The navigator
 * relays tree positions, and every widget joined to it reads an incoming
 * {@code NavigateTo} as "a syllable was selected". A character is not a tree
 * position — terminals hold characters as <em>content</em>, not as children —
 * so putting character selections on that bus would have the syllable panes
 * treat a codepoint as a path and misbehave.</p>
 *
 * <p>Just a redirect, like the navigator: an incoming {@code ZiSelected} is
 * rebroadcast to every member as {@code ShowZi}. The producer never names its
 * consumer.</p>
 *
 * <h2>State shape</h2>
 * <pre>{@code
 * {
 *     lastSelected : { codePoint, glyph } | null,
 *     recentUnknown: [{ kind, from }]     // bounded at 10
 * }
 * }</pre>
 */
public record ZiSelectionSecretaryModule() implements DomModule<ZiSelectionSecretaryModule> {

    /** The single export — a JS object with {@code initial} and {@code behavior}. */
    public record ZiSelectionSecretary()
            implements Exportable._Constant<ZiSelectionSecretaryModule> {}

    public static final ZiSelectionSecretaryModule INSTANCE = new ZiSelectionSecretaryModule();

    @Override
    public ImportsFor<ZiSelectionSecretaryModule> imports() {
        return ImportsFor.<ZiSelectionSecretaryModule>builder().build();
    }

    @Override
    public ExportsOf<ZiSelectionSecretaryModule> exports() {
        return new ExportsOf<>(INSTANCE, List.of(new ZiSelectionSecretary()));
    }
}

File: kranji-reading/kranji-reading-app/src/main/resources/homing/js/kranji/reading/app/zi/ZiSelectionSecretaryModule.js
// =============================================================================
// ZiSelectionSecretaryModule — the Secretary for the character-selection
// Party.
//
// A second bus, separate from navigation on purpose. The navigator relays
// tree positions; a character is not a tree position (terminals hold
// characters as content, not as children), so a codepoint travelling on that
// bus would be read as a path by the syllable panes.
//
// Just a redirect: an incoming ZiSelected is rebroadcast to every member as
// ShowZi. Pure (state, envelope) -> Step; no DOM, no console, no captures.
// =============================================================================

var ZiSelectionSecretary = {

    initial: {
        lastSelected:  null,
        recentUnknown: []
    },

    /**
     * Pure routing function. (state, envelope) -> { newState, actions }.
     */
    behavior: function (state, envelope) {
        var msg = envelope.message;

        if (msg.kind === "ZiSelected") {
            return {
                newState: {
                    lastSelected:  msg.zi,
                    recentUnknown: state.recentUnknown
                },
                actions: [{
                    kind:    "BroadcastToMembers",
                    message: { kind: "ShowZi", zi: msg.zi }
                }]
            };
        }

        // Anything else is remembered for observability and otherwise ignored -
        // a bus that throws on an unknown kind couples every producer to every
        // consumer's vocabulary.
        var seen = state.recentUnknown.concat([
            { kind: msg.kind, from: envelope.from }
        ]);
        return {
            newState: {
                lastSelected:  state.lastSelected,
                recentUnknown: seen.slice(-10)
            },
            actions: []
        };
    }
};
